File: skills/docforge/_shared/runtime/common/python/provenance_store.py
#!/usr/bin/env python3
"""Folder-mirrored JSON sidecar store for Docforge document metadata.

When `project.provenance_storage` is `json` (the default), the public
frontmatter identity (id, title, description) and the private
`docforge_provenance` object live in one git-tracked JSON file per docs
folder under `.docforge/provenance/`, and the markdown files carry no
frontmatter at all. With `markdown` storage the legacy inline frontmatter
layout is kept. Both runtimes share this module so file moves, reads, and
writes stay byte-identical.
"""

from __future__ import annotations

import json
from pathlib import Path, PurePosixPath
from typing import Any

from runtime.common.python.provenance_frontmatter import (
    emit_document_frontmatter,
    parse_yaml_mapping,
    split_frontmatter,
)

SIDECAR_SCHEMA = "1.0"
STORAGE_JSON = "json"
STORAGE_MARKDOWN = "markdown"
SIDECAR_DIRNAME = "provenance"
PUBLIC_FIELDS = ("id", "title", "description")


class SidecarError(ValueError):
    """Raised when a sidecar file cannot be read or written."""


def storage_for(manifest: dict) -> str:
    """The project's provenance storage mode; json when unset (default)."""
    value = (manifest.get("project") or {}).get("provenance_storage")
    return value if value in {STORAGE_JSON, STORAGE_MARKDOWN} else STORAGE_JSON


def sidecar_root(repo: Path) -> Path:
    return repo / ".docforge" / SIDECAR_DIRNAME


def folder_of(doc_path: str) -> str:
    """POSIX folder path for a document; "" for repo-root documents."""
    parent = PurePosixPath(doc_path.replace("\\", "/")).parent
    folder = parent.as_posix()
    return "" if folder in ("", ".") else folder


def sidecar_path(repo: Path, folder: str) -> Path:
    """`.docforge/provenance/<folder>.json`; `root.json` for repo root."""
    name = f"{folder}.json" if folder else "root.json"
    return sidecar_root(repo) / name


def read_sidecar(repo: Path, folder: str) -> dict | None:
    """The sidecar object for a folder, or None when absent."""
    path = sidecar_path(repo, folder)
    if not path.is_file():
        return None
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise SidecarError(f"invalid sidecar {path.relative_to(repo)}: {exc}") from exc
    if not isinstance(data, dict) or not isinstance(data.get("files"), dict):
        raise SidecarError(f"invalid sidecar shape: {path.relative_to(repo)}")
    return data


def write_sidecar(repo: Path, folder: str, data: dict) -> None:
    """Write a folder sidecar, creating `.docforge/provenance/` as needed."""
    path = sidecar_path(repo, folder)
    files = data.get("files") if isinstance(data, dict) else None
    if not isinstance(files, dict) or not files:
        if path.is_file():
            path.unlink()
        return
    path.parent.mkdir(parents=True, exist_ok=True)
    payload = {"schema": SIDECAR_SCHEMA, "folder": folder, "files": files}
    path.write_text(
        json.dumps(payload, indent=2, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )


def file_name_of(doc_path: str) -> str:
    return PurePosixPath(doc_path.replace("\\", "/")).name


def entry_for(repo: Path, doc_path: str) -> dict | None:
    """The sidecar entry for one document, or None."""
    data = read_sidecar(repo, folder_of(doc_path))
    if data is None:
        return None
    entry = data["files"].get(file_name_of(doc_path))
    return entry if isinstance(entry, dict) else None


def write_entry(repo: Path, doc_path: str, entry: dict) -> None:
    """Merge one document's entry into its folder sidecar."""
    folder = folder_of(doc_path)
    name = file_name_of(doc_path)
    data = read_sidecar(repo, folder)
    files = dict(data["files"]) if data else {}
    files[name] = entry
    write_sidecar(repo, folder, {"files": files})


def remove_entry(repo: Path, doc_path: str) -> None:
    """Drop one document's entry, deleting the sidecar when it becomes empty."""
    folder = folder_of(doc_path)
    name = file_name_of(doc_path)
    data = read_sidecar(repo, folder)
    if not data or name not in data["files"]:
        return
    files = dict(data["files"])
    del files[name]
    write_sidecar(repo, folder, {"files": files})


def read_inline(text: str) -> tuple[str, dict | None]:
    """Parse a markdown file's full frontmatter mapping (id/title/description
    plus docforge_provenance) without interpreting schema state."""
    raw, _body, _end = split_frontmatter(text)
    if raw is None:
        return "missing", None
    try:
        data = parse_yaml_mapping(raw)
    except ValueError:
        return "unparseable", None
    if not isinstance(data, dict):
        return "missing", None
    return "ok", data


def read_doc_metadata(repo: Path, doc: dict, storage: str) -> dict:
    """Mode-aware read of one document's metadata.

    Returns {"state", "public", "provenance", "source"} where state is one
    of ok | missing | unparseable | inline and source is sidecar|markdown.
    In json mode a file whose inline frontmatter has not been migrated yet
    reports state `inline` so callers can trigger the move.
    """
    doc_path = doc.get("path", "")
    target = repo / doc_path
    if storage == STORAGE_JSON:
        entry = entry_for(repo, doc_path)
        if isinstance(entry, dict) and isinstance(entry.get("provenance"), dict):
            return {
                "state": "ok",
                "public": {key: entry.get(key) for key in PUBLIC_FIELDS},
                "provenance": entry["provenance"],
                "source": "sidecar",
            }
        if target.is_file():
            text = target.read_text(encoding="utf-8", errors="replace")
            state, data = read_inline(text)
            if state == "ok" and isinstance(data.get("docforge_provenance"), dict):
                return {
                    "state": "inline",
                    "public": {key: data.get(key) for key in PUBLIC_FIELDS},
                    "provenance": data["docforge_provenance"],
                    "source": "markdown",
                }
            if state == "unparseable":
                return {"state": "unparseable", "public": None, "provenance": None, "source": "markdown"}
        return {"state": "missing", "public": None, "provenance": None, "source": "sidecar"}
    if not target.is_file():
        return {"state": "missing", "public": None, "provenance": None, "source": "markdown"}
    text = target.read_text(encoding="utf-8", errors="replace")
    state, data = read_inline(text)
    if state == "unparseable":
        return {"state": "unparseable", "public": None, "provenance": None, "source": "markdown"}
    if state == "ok" and isinstance(data.get("docforge_provenance"), dict):
        return {
            "state": "ok",
            "public": {key: data.get(key) for key in PUBLIC_FIELDS},
            "provenance": data["docforge_provenance"],
            "source": "markdown",
        }
    return {"state": "missing", "public": None, "provenance": None, "source": "markdown"}


def public_from_manifest(doc: dict) -> dict:
    """Public identity (id/title/description) seeded from the manifest entry."""
    title = doc.get("title") or doc["id"].replace("-", " ").replace("_", " ").title()
    public: dict[str, Any] = {"id": doc["id"], "title": title}
    if doc.get("description"):
        public["description"] = doc["description"]
    return public


def move_inline_to_sidecar(repo: Path, doc: dict, storage: str) -> str:
    """Move a document's inline frontmatter into the folder sidecar and strip
    it from the markdown. No-op unless json storage and inline exists."""
    if storage != STORAGE_JSON:
        return "skip"
    doc_path = doc.get("path", "")
    target = repo / doc_path
    if not target.is_file():
        return "missing"
    text = target.read_text(encoding="utf-8", errors="replace")
    state, data = read_inline(text)
    if state == "unparseable":
        return "unparseable"
    if state == "missing" or not isinstance(data.get("docforge_provenance"), dict):
        return "no-frontmatter"
    public = {key: data.get(key) for key in PUBLIC_FIELDS if data.get(key)}
    if not public.get("id"):
        public["id"] = doc.get("id", "")
    if not public.get("title"):
        public["title"] = doc.get("title") or doc.get("id", "document").replace("-", " ").title()
    if not public.get("description") and doc.get("description"):
        public["description"] = doc["description"]
    entry = {key: value for key, value in public.items() if value}
    entry["provenance"] = data["docforge_provenance"]
    write_entry(repo, doc_path, entry)
    _raw, body, _end = split_frontmatter(text)
    target.write_text(body.lstrip("\n"), encoding="utf-8")
    return "moved"


def move_sidecar_to_inline(repo: Path, doc: dict) -> str:
    """Re-emit a sidecar entry as inline frontmatter and drop the entry."""
    doc_path = doc.get("path", "")
    entry = entry_for(repo, doc_path)
    if not isinstance(entry, dict):
        return "no-sidecar"
    public = {key: entry.get(key) for key in PUBLIC_FIELDS if entry.get(key)}
    title = public.get("title") or doc.get("title") or doc.get("id", "document")
    doc_id = public.get("id") or doc.get("id", "")
    description = public.get("description")
    frontmatter = emit_document_frontmatter(
        doc_id, title, entry["provenance"], description
    )
    target = repo / doc_path
    if target.is_file():
        text = target.read_text(encoding="utf-8", errors="replace")
        _raw, body, _end = split_frontmatter(text)
        content = frontmatter + body.lstrip("\n")
    else:
        content = frontmatter
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(content, encoding="utf-8")
    remove_entry(repo, doc_path)
    return "moved"

File: skills/docforge/_shared/runtime/common/js/provenance_store.js
#!/usr/bin/env node
"use strict";
/* Folder-mirrored JSON sidecar store for Docforge document metadata.
 * With `project.provenance_storage: json` (default) the public frontmatter
 * identity (id, title, description) and `docforge_provenance` live in one
 * git-tracked JSON file per docs folder under `.docforge/provenance/`; the
 * markdown files carry no frontmatter. With `markdown` storage the legacy
 * inline layout is kept. Mirrors common/python/provenance_store.py.
 */

const fs = require("fs");
const path = require("path");
const pf = require("./provenance_frontmatter.js");

const SIDECAR_SCHEMA = "1.0";
const STORAGE_JSON = "json";
const STORAGE_MARKDOWN = "markdown";
const SIDECAR_DIRNAME = "provenance";
const PUBLIC_FIELDS = ["id", "title", "description"];

class SidecarError extends Error {
  constructor(message) {
    super(message);
    this.name = "SidecarError";
  }
}

function storageFor(manifest) {
  const value = (manifest && manifest.project && manifest.project.provenance_storage) || null;
  return value === STORAGE_JSON || value === STORAGE_MARKDOWN ? value : STORAGE_JSON;
}

function sidecarRoot(repo) {
  return path.join(repo, ".docforge", SIDECAR_DIRNAME);
}

function folderOf(docPath) {
  const normalized = String(docPath).replace(/\\/g, "/");
  const parent = normalized.includes("/") ? normalized.slice(0, normalized.lastIndexOf("/")) : "";
  return parent === "." ? "" : parent;
}

function sidecarPath(repo, folder) {
  const name = folder ? `${folder}.json` : "root.json";
  return path.join(sidecarRoot(repo), name);
}

function readSidecar(repo, folder) {
  const target = sidecarPath(repo, folder);
  if (!fs.existsSync(target)) return null;
  let data;
  try {
    data = JSON.parse(fs.readFileSync(target, "utf8"));
  } catch (error) {
    throw new SidecarError(`invalid sidecar ${path.relative(repo, target)}: ${error.message}`);
  }
  if (!data || typeof data !== "object" || Array.isArray(data)
    || !data.files || typeof data.files !== "object" || Array.isArray(data.files)) {
    throw new SidecarError(`invalid sidecar shape: ${path.relative(repo, target)}`);
  }
  return data;
}

function writeSidecar(repo, folder, data) {
  const target = sidecarPath(repo, folder);
  const files = data && typeof data === "object" ? data.files : null;
  if (!files || typeof files !== "object" || Array.isArray(files) || Object.keys(files).length === 0) {
    if (fs.existsSync(target)) fs.unlinkSync(target);
    return;
  }
  fs.mkdirSync(path.dirname(target), { recursive: true });
  const payload = { schema: SIDECAR_SCHEMA, folder, files };
  fs.writeFileSync(target, JSON.stringify(payload, null, 2) + "\n", "utf8");
}

function fileNameOf(docPath) {
  const normalized = String(docPath).replace(/\\/g, "/");
  return normalized.includes("/") ? normalized.slice(normalized.lastIndexOf("/") + 1) : normalized;
}

function entryFor(repo, docPath) {
  const data = readSidecar(repo, folderOf(docPath));
  if (!data) return null;
  const entry = data.files[fileNameOf(docPath)];
  return entry && typeof entry === "object" && !Array.isArray(entry) ? entry : null;
}

function writeEntry(repo, docPath, entry) {
  const folder = folderOf(docPath);
  const name = fileNameOf(docPath);
  const data = readSidecar(repo, folder);
  const files = data ? { ...data.files } : {};
  files[name] = entry;
  writeSidecar(repo, folder, { files });
}

function removeEntry(repo, docPath) {
  const folder = folderOf(docPath);
  const name = fileNameOf(docPath);
  const data = readSidecar(repo, folder);
  if (!data || !(name in data.files)) return;
  const files = { ...data.files };
  delete files[name];
  writeSidecar(repo, folder, { files });
}

function readInline(text) {
  const { raw } = pf.splitFrontmatter(text);
  if (raw == null) return { state: "missing", data: null };
  let data;
  try {
    data = pf.parseYamlMapping(raw);
  } catch {
    return { state: "unparseable", data: null };
  }
  if (!data || typeof data !== "object" || Array.isArray(data)) {
    return { state: "missing", data: null };
  }
  return { state: "ok", data };
}

function readDocMetadata(repo, doc, storage) {
  const docPath = (doc && doc.path) || "";
  const target = path.join(repo, docPath);
  if (storage === STORAGE_JSON) {
    const entry = entryFor(repo, docPath);
    if (entry && entry.provenance && typeof entry.provenance === "object") {
      return {
        state: "ok",
        public: publicOf(entry),
        provenance: entry.provenance,
        source: "sidecar",
      };
    }
    if (fs.existsSync(target)) {
      const text = fs.readFileSync(target, "utf8");
      const { state, data } = readInline(text);
      if (state === "ok" && data.docforge_provenance && typeof data.docforge_provenance === "object") {
        return {
          state: "inline",
          public: publicOf(data),
          provenance: data.docforge_provenance,
          source: "markdown",
        };
      }
      if (state === "unparseable") {
        return { state: "unparseable", public: null, provenance: null, source: "markdown" };
      }
    }
    return { state: "missing", public: null, provenance: null, source: "sidecar" };
  }
  if (!fs.existsSync(target)) {
    return { state: "missing", public: null, provenance: null, source: "markdown" };
  }
  const text = fs.readFileSync(target, "utf8");
  const { state, data } = readInline(text);
  if (state === "unparseable") {
    return { state: "unparseable", public: null, provenance: null, source: "markdown" };
  }
  if (state === "ok" && data.docforge_provenance && typeof data.docforge_provenance === "object") {
    return {
      state: "ok",
      public: publicOf(data),
      provenance: data.docforge_provenance,
      source: "markdown",
    };
  }
  return { state: "missing", public: null, provenance: null, source: "markdown" };
}

function publicOf(data) {
  const publicMeta = {};
  for (const key of PUBLIC_FIELDS) {
    if (data && data[key]) publicMeta[key] = data[key];
  }
  return publicMeta;
}

function publicFromManifest(doc) {
  const title = doc.title || doc.id.replace(/-/g, " ").replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
  const publicMeta = { id: doc.id, title };
  if (doc.description) publicMeta.description = doc.description;
  return publicMeta;
}

function moveInlineToSidecar(repo, doc, storage) {
  if (storage !== STORAGE_JSON) return "skip";
  const docPath = (doc && doc.path) || "";
  const target = path.join(repo, docPath);
  if (!fs.existsSync(target)) return "missing";
  const text = fs.readFileSync(target, "utf8");
  const { state, data } = readInline(text);
  if (state === "unparseable") return "unparseable";
  if (state === "missing" || !data.docforge_provenance || typeof data.docforge_provenance !== "object") {
    return "no-frontmatter";
  }
  const publicMeta = publicOf(data);
  if (!publicMeta.id) publicMeta.id = doc.id || "";
  if (!publicMeta.title) {
    publicMeta.title = doc.title || (doc.id || "document").replace(/-/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
  }
  if (!publicMeta.description && doc.description) publicMeta.description = doc.description;
  const entry = { ...publicMeta, provenance: data.docforge_provenance };
  writeEntry(repo, docPath, entry);
  const { body } = pf.splitFrontmatter(text);
  fs.writeFileSync(target, body.replace(/^\n+/, ""), "utf8");
  return "moved";
}

function moveSidecarToInline(repo, doc) {
  const docPath = (doc && doc.path) || "";
  const entry = entryFor(repo, docPath);
  if (!entry) return "no-sidecar";
  const publicMeta = publicOf(entry);
  const title = publicMeta.title || doc.title || doc.id || "document";
  const docId = publicMeta.id || doc.id || "";
  const frontmatter = pf.emitDocumentFrontmatter(docId, title, entry.provenance, publicMeta.description || null);
  const target = path.join(repo, docPath);
  let content = frontmatter;
  if (fs.existsSync(target)) {
    const text = fs.readFileSync(target, "utf8");
    const { body } = pf.splitFrontmatter(text);
    content = frontmatter + body.replace(/^\n+/, "");
  }
  fs.mkdirSync(path.dirname(target), { recursive: true });
  fs.writeFileSync(target, content, "utf8");
  removeEntry(repo, docPath);
  return "moved";
}

module.exports = {
  SIDECAR_SCHEMA,
  STORAGE_JSON,
  STORAGE_MARKDOWN,
  SIDECAR_DIRNAME,
  PUBLIC_FIELDS,
  SidecarError,
  storageFor,
  sidecarRoot,
  folderOf,
  sidecarPath,
  readSidecar,
  writeSidecar,
  fileNameOf,
  entryFor,
  writeEntry,
  removeEntry,
  readInline,
  readDocMetadata,
  publicOf,
  publicFromManifest,
  moveInlineToSidecar,
  moveSidecarToInline,
};
